samza-core/src/main/java/org/apache/samza/container/grouper/stream/SSPGrouperProxy.java

```diff
@@ -0,0 +1,195 @@
+    this.grouper = grouper;
+    this.broadcastSystemStreamPartitions = new TaskConfigJava(config).getBroadcastSystemStreamPartitions();
+    this.streamPartitionMapper = getStreamPartitionMapper(config);
+  }
+
+  /**
+   * 1. Invokes the sub-class {@link org.apache.samza.container.grouper.task.TaskNameGrouper#group(Set)} method to generate the task to partition assignments.
+   * 2. Uses the {@link StreamPartitionMapper} to compute the previous {@link SystemStreamPartition} for every partition of the expanded streams.
+   * 3. Uses the previous, current task to partition assignments and result of {@link StreamPartitionMapper} to redistribute the expanded {@link SystemStreamPartition}'s
+   * to correct tasks after the stream expansion or contraction.
+   * @param ssps the input system stream partitions of the job.
+   * @param grouperContext the grouper context holding metadata for the previous job execution.
+   * @return the grouped {@link TaskName} to {@link SystemStreamPartition} assignments.
+   */
+  public Map<TaskName, Set<SystemStreamPartition>> group(Set<SystemStreamPartition> ssps, GrouperContext grouperContext) {
+    Map<TaskName, Set<SystemStreamPartition>> currentTaskAssignments = grouper.group(ssps);
+
+    if (grouperContext.getPreviousTaskToSSPAssignment().isEmpty()) {
+      LOGGER.info("Previous task to partition assignment does not exist. Using the result from the group method.");
+      return currentTaskAssignments;
+    }
+
+    Map<SystemStreamPartition, TaskName> previousSSPToTask = getPreviousSSPToTaskMapping(grouperContext);
+
+    Map<TaskName, PartitionGroup> taskToPartitionGroup = new HashMap<>();
+    currentTaskAssignments.forEach((taskName, systemStreamPartitions) -> taskToPartitionGroup.put(taskName, new PartitionGroup(taskName, systemStreamPartitions)));
+
+    Map<SystemStream, Integer> previousStreamToPartitionCount = getStreamToPartitionCount(grouperContext.getPreviousTaskToSSPAssignment());
+    Map<SystemStream, Integer> currentStreamToPartitionCount = getStreamToPartitionCount(currentTaskAssignments);
+
+    for (Map.Entry<TaskName, Set<SystemStreamPartition>> entry : currentTaskAssignments.entrySet()) {
+      TaskName currentlyAssignedTask = entry.getKey();
+      for (SystemStreamPartition currentSystemStreamPartition : entry.getValue()) {
+        if (broadcastSystemStreamPartitions.contains(currentSystemStreamPartition)) {
+          LOGGER.info("SystemStreamPartition: {} is part of broadcast stream. Skipping reassignment.", currentSystemStreamPartition);
+        } else {
+          SystemStream systemStream = currentSystemStreamPartition.getSystemStream();
+
+          Integer previousStreamPartitionCount = previousStreamToPartitionCount.getOrDefault(systemStream, 0);
+          Integer currentStreamPartitionCount = currentStreamToPartitionCount.getOrDefault(systemStream, 0);
+
+          if (previousStreamPartitionCount > 0 && currentStreamPartitionCount % previousStreamPartitionCount == 0) {
```

Comment:
Maybe this check for "previous count needing to be a multiplier of the current partition count" should be enforced only by the `StreamPartitionMapper`? Here, if the condition is false, then it loosely fails by logging a warning and will never hit the same Precondition check in `HashStreamPartitionMapper#getSSPAfterPartitionChange`. Might be cleaner to consolidate the logic.